File: day_27_set_operations/day_27_set_operations.js
"use strict";

/*
 * SET OPERATIONS IN JAVASCRIPT
 * =============================
 *
 * Self-contained executable study file.
 * Run with:
 *     node set_operations.js
 *
 * JavaScript Set is a collection of unique values. It is useful for
 * membership testing, deduplication, set-like operations, graph traversal,
 * permissions, data reconciliation, and indexing.
 */

// ============================================================================
// 1. FUNDAMENTALS
// ============================================================================

function fundamentals() {
    console.log("\n" + "=".repeat(78));
    console.log("1. FUNDAMENTALS");
    console.log("=".repeat(78));

    const numbers = new Set([1, 2, 3, 3, 2, 1]);

    console.log("Unique values:", numbers);
    console.log("Size:", numbers.size);

    // JavaScript Set uses has() for membership testing.
    console.log("Has 2:", numbers.has(2));
    console.log("Has 99:", numbers.has(99));

    // Set has no index-based access. Convert to an array when positional
    // access is actually required.
    console.log("First element through iterator:", numbers.values().next().value);

    const emptySet = new Set();
    console.log("Empty set:", emptySet);

    // JavaScript Set can contain objects because object identity is used.
    const firstObject = { id: 1 };
    const secondObject = { id: 1 };
    const objectSet = new Set([firstObject, secondObject]);

    console.log("Two different objects:", objectSet.size);
    console.log("Same object reference:", objectSet.has(firstObject));
}

// ============================================================================
// 2. ADDING, DELETING, AND CLEARING
// ============================================================================

function mutationOperations() {
    console.log("\n" + "=".repeat(78));
    console.log("2. MUTATION OPERATIONS");
    console.log("=".repeat(78));

    const values = new Set([1, 2, 3]);

    values.add(4);
    values.add(4); // Adding an existing value does not create a duplicate.

    console.log("After add:", values);

    values.delete(2);
    console.log("After delete:", values);

    // delete() returns true when an element existed and was removed.
    console.log("Delete missing element:", values.delete(999));

    values.clear();
    console.log("After clear:", values);
}

// ============================================================================
// 3. ITERATION
// ============================================================================

function iterationExamples() {
    console.log("\n" + "=".repeat(78));
    console.log("3. ITERATION");
    console.log("=".repeat(78));

    const values = new Set(["alpha", "beta", "gamma"]);

    for (const value of values) {
        console.log("for...of:", value);
    }

    values.forEach((value) => {
        console.log("forEach:", value);
    });

    console.log("Array conversion:", [...values]);
}

// ============================================================================
// 4. SET-LIKE MATHEMATICAL OPERATIONS
// ============================================================================

function union(first, second) {
    return new Set([...first, ...second]);
}

function intersection(first, second) {
    // Iterate through the smaller set when possible.
    const [smaller, larger] =
        first.size <= second.size ? [first, second] : [second, first];

    return new Set([...smaller].filter((value) => larger.has(value)));
}

function difference(first, second) {
    return new Set([...first].filter((value) => !second.has(value)));
}

function symmetricDifference(first, second) {
    return new Set([
        ...[...first].filter((value) => !second.has(value)),
        ...[...second].filter((value) => !first.has(value)),
    ]);
}

function mathematicalOperations() {
    console.log("\n" + "=".repeat(78));
    console.log("4. MATHEMATICAL SET OPERATIONS");
    console.log("=".repeat(78));

    const A = new Set([1, 2, 3, 4]);
    const B = new Set([3, 4, 5, 6]);

    console.log("A:", A);
    console.log("B:", B);
    console.log("Union:", union(A, B));
    console.log("Intersection:", intersection(A, B));
    console.log("A - B:", difference(A, B));
    console.log("B - A:", difference(B, A));
    console.log("Symmetric difference:", symmetricDifference(A, B));
}

// ============================================================================
// 5. SUBSET, SUPERSET, AND DISJOINTNESS
// ============================================================================

function isSubset(subset, superset) {
    for (const value of subset) {
        if (!superset.has(value)) {
            return false;
        }
    }
    return true;
}

function isDisjoint(first, second) {
    const [smaller, larger] =
        first.size <= second.size ? [first, second] : [second, first];

    for (const value of smaller) {
        if (larger.has(value)) {
            return false;
        }
    }

    return true;
}

function relationExamples() {
    console.log("\n" + "=".repeat(78));
    console.log("5. SET RELATIONS");
    console.log("=".repeat(78));

    const A = new Set([1, 2]);
    const B = new Set([1, 2, 3]);
    const C = new Set([8, 9]);

    console.log("A subset of B:", isSubset(A, B));
    console.log("B subset of A:", isSubset(B, A));
    console.log("A and C disjoint:", isDisjoint(A, C));
}

// ============================================================================
// 6. DEDUPLICATION
// ============================================================================

function deduplicationExamples() {
    console.log("\n" + "=".repeat(78));
    console.log("6. DEDUPLICATION");
    console.log("=".repeat(78));

    const input = [4, 2, 4, 1, 2, 3, 1];

    const unique = [...new Set(input)];

    console.log("Input:", input);
    console.log("Unique values:", unique);

    const emails = [
        "Alice@example.com",
        "alice@example.com",
        "bob@example.com",
        " BOB@example.com ",
    ];

    // Normalization is required when values that differ only in formatting
    // should be treated as the same logical entity.
    const normalizedEmails = new Set(
        emails.map((email) => email.trim().toLowerCase())
    );

    console.log("Normalized unique emails:", [...normalizedEmails]);
}

// ============================================================================
// 7. EDGE CASES AND JAVASCRIPT EQUALITY
// ============================================================================

function edgeCases() {
    console.log("\n" + "=".repeat(78));
    console.log("7. EDGE CASES AND EQUALITY");
    console.log("=".repeat(78));

    // Set uses SameValueZero-style equality:
    // NaN matches NaN, and +0 and -0 are considered the same value.
    const specialValues = new Set([NaN, NaN, +0, -0]);

    console.log("Special value count:", specialValues.size);
    console.log("Contains NaN:", specialValues.has(NaN));
    console.log("Contains -0:", specialValues.has(-0));

    // Objects are compared by reference, not by their contents.
    const objectA = { name: "Atul" };
    const objectB = { name: "Atul" };
    const objects = new Set([objectA, objectB]);

    console.log("Equal-looking objects stored separately:", objects.size);
    console.log("objectA found:", objects.has(objectA));
    console.log("New equivalent object found:", objects.has({ name: "Atul" }));
}

// ============================================================================
// 8. MULTISET LIMITATION
// ============================================================================

function frequencyMap(values) {
    const counts = new Map();

    for (const value of values) {
        counts.set(value, (counts.get(value) ?? 0) + 1);
    }

    return counts;
}

function multisetDifference(first, second) {
    const firstCounts = frequencyMap(first);
    const secondCounts = frequencyMap(second);
    const result = [];

    for (const [value, count] of firstCounts) {
        const remaining = count - (secondCounts.get(value) ?? 0);

        for (let i = 0; i < Math.max(0, remaining); i += 1) {
            result.push(value);
        }
    }

    return result;
}

function multisetExample() {
    console.log("\n" + "=".repeat(78));
    console.log("8. MULTISET LIMITATION");
    console.log("=".repeat(78));

    const first = ["a", "a", "a", "b", "c"];
    const second = ["a", "b"];

    console.log("Ordinary Set:", new Set(first));
    console.log("Frequency map:", frequencyMap(first));
    console.log("Multiset difference:", multisetDifference(first, second));
}

// ============================================================================
// 9. PERMISSIONS SYSTEM
// ============================================================================

function permissionSystem() {
    console.log("\n" + "=".repeat(78));
    console.log("9. PERMISSION SYSTEM");
    console.log("=".repeat(78));

    const required = new Set(["read", "write"]);
    const granted = new Set(["read", "write", "analytics"]);

    const authorized = isSubset(required, granted);
    const extra = difference(granted, required);
    const missing = difference(required, granted);

    console.log("Authorized:", authorized);
    console.log("Extra permissions:", extra);
    console.log("Missing permissions:", missing);
}

// ============================================================================
// 10. SEARCH INDEX
// ============================================================================

function buildInvertedIndex(documents) {
    const index = new Map();

    for (const [documentId, text] of Object.entries(documents)) {
        const words = new Set(
            text
                .toLowerCase()
                .split(/\s+/)
                .map((word) => word.replace(/^[^\p{L}\p{N}]+|[^\p{L}\p{N}]+$/gu, ""))
                .filter(Boolean)
        );

        for (const word of words) {
            if (!index.has(word)) {
                index.set(word, new Set());
            }

            index.get(word).add(documentId);
        }
    }

    return index;
}

function searchIndex(index, requiredTerms) {
    if (requiredTerms.length === 0) {
        return new Set();
    }

    const normalizedTerms = requiredTerms.map((term) => term.toLowerCase());

    let result = new Set(index.get(normalizedTerms[0]) ?? []);

    for (const term of normalizedTerms.slice(1)) {
        result = intersection(result, index.get(term) ?? new Set());
    }

    return result;
}

function searchExample() {
    console.log("\n" + "=".repeat(78));
    console.log("10. INVERTED INDEX");
    console.log("=".repeat(78));

    const documents = {
        doc1: "JavaScript supports Set operations and Map structures.",
        doc2: "Set operations are useful for data processing.",
        doc3: "JavaScript data structures include arrays sets and maps.",
    };

    const index = buildInvertedIndex(documents);

    console.log("Documents containing set:", index.get("set"));
    console.log(
        "Documents containing set and operations:",
        searchIndex(index, ["set", "operations"])
    );
}

// ============================================================================
// 11. GRAPH TRAVERSAL
// ============================================================================

function breadthFirstSearch(graph, start) {
    const visited = new Set();
    const queue = [start];
    let index = 0;

    while (index < queue.length) {
        const vertex = queue[index];
        index += 1;

        if (visited.has(vertex)) {
            continue;
        }

        visited.add(vertex);

        for (const neighbor of graph.get(vertex) ?? []) {
            if (!visited.has(neighbor)) {
                queue.push(neighbor);
            }
        }
    }

    return visited;
}

function graphExample() {
    console.log("\n" + "=".repeat(78));
    console.log("11. GRAPH APPLICATION");
    console.log("=".repeat(78));

    const graph = new Map([
        ["A", new Set(["B", "C"])],
        ["B", new Set(["A", "C", "D"])],
        ["C", new Set(["A", "B"])],
        ["D", new Set(["B"])],
    ]);

    console.log("Common neighbors A/B:", intersection(graph.get("A"), graph.get("B")));
    console.log("Reachable from A:", breadthFirstSearch(graph, "A"));
}

// ============================================================================
// 12. DATA RECONCILIATION
// ============================================================================

function reconciliation() {
    console.log("\n" + "=".repeat(78));
    console.log("12. DATA RECONCILIATION");
    console.log("=".repeat(78));

    const databaseIds = new Set([101, 102, 103, 104, 105]);
    const apiIds = new Set([103, 104, 105, 106, 107]);

    console.log("Common:", intersection(databaseIds, apiIds));
    console.log("Only database:", difference(databaseIds, apiIds));
    console.log("Only API:", difference(apiIds, databaseIds));
    console.log("All IDs:", union(databaseIds, apiIds));
}

// ============================================================================
// 13. ASYNCHRONOUS SET PROCESSING
// ============================================================================

function delay(milliseconds) {
    return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

async function asynchronousProcessing() {
    console.log("\n" + "=".repeat(78));
    console.log("13. ASYNCHRONOUS PROCESSING");
    console.log("=".repeat(78));

    const pendingJobs = new Set(["job-101", "job-102", "job-103"]);

    console.log("Pending jobs:", pendingJobs);

    // The Set tracks state while asynchronous work occurs.
    for (const job of [...pendingJobs]) {
        await delay(5);
        pendingJobs.delete(job);
        console.log(`Completed ${job}; remaining:`, pendingJobs.size);
    }
}

// ============================================================================
// 14. PERFORMANCE AND DESIGN NOTES
// ============================================================================

function performanceNotes() {
    console.log("\n" + "=".repeat(78));
    console.log("14. PERFORMANCE AND DESIGN NOTES");
    console.log("=".repeat(78));

    const largeArray = Array.from({ length: 100_000 }, (_, index) => index);
    const largeSet = new Set(largeArray);
    const target = 99_999;

    console.time("Array membership");
    largeArray.includes(target);
    console.timeEnd("Array membership");

    console.time("Set membership");
    largeSet.has(target);
    console.timeEnd("Set membership");

    console.log("Typical Set membership is average O(1).");
    console.log("Array includes() is O(n) in the general case.");
    console.log("Set construction costs O(n) for n input elements.");
    console.log("Set consumes memory for its hash/collection structure.");
}

// ============================================================================
// 15. ERROR HANDLING AND VALIDATION
// ============================================================================

function validateSet(value, name = "value") {
    if (!(value instanceof Set)) {
        throw new TypeError(`${name} must be a Set`);
    }

    return true;
}

function safeIntersection(first, second) {
    validateSet(first, "first");
    validateSet(second, "second");
    return intersection(first, second);
}

function validationExample() {
    console.log("\n" + "=".repeat(78));
    console.log("15. ERROR HANDLING AND VALIDATION");
    console.log("=".repeat(78));

    try {
        safeIntersection(new Set([1, 2]), [2, 3]);
    } catch (error) {
        console.log("Validation error:", error.message);
    }

    console.log(
        "Valid intersection:",
        safeIntersection(new Set([1, 2]), new Set([2, 3]))
    );
}

// ============================================================================
// 16. TESTS
// ============================================================================

function assert(condition, message) {
    if (!condition) {
        throw new Error(`Assertion failed: ${message}`);
    }
}

function runTests() {
    console.log("\n" + "=".repeat(78));
    console.log("16. SELF-TESTS");
    console.log("=".repeat(78));

    const A = new Set([1, 2, 3]);
    const B = new Set([3, 4, 5]);

    assert(
        [...union(A, B)].sort((a, b) => a - b).join(",") === "1,2,3,4,5",
        "union"
    );

    assert(
        [...intersection(A, B)].join(",") === "3",
        "intersection"
    );

    assert(
        [...difference(A, B)].sort((a, b) => a - b).join(",") === "1,2",
        "difference"
    );

    assert(
        [...symmetricDifference(A, B)]
            .sort((a, b) => a - b)
            .join(",") === "1,2,4,5",
        "symmetric difference"
    );

    assert(isSubset(new Set([1, 2]), A), "subset");
    assert(isDisjoint(A, new Set([8, 9])), "disjointness");

    assert(
        JSON.stringify([...new Set([1, 2, 1, 3])]) === JSON.stringify([1, 2, 3]),
        "deduplication"
    );

    assert(
        [...new Set([NaN, NaN])].length === 1,
        "NaN uniqueness"
    );

    console.log("All tests passed.");
}

// ============================================================================
// 17. MAIN
// ============================================================================

async function main() {
    console.log("=".repeat(78));
    console.log("SET OPERATIONS IN JAVASCRIPT: BEGINNER TO ADVANCED");
    console.log("=".repeat(78));

    fundamentals();
    mutationOperations();
    iterationExamples();
    mathematicalOperations();
    relationExamples();
    deduplicationExamples();
    edgeCases();
    multisetExample();
    permissionSystem();
    searchExample();
    graphExample();
    reconciliation();
    await asynchronousProcessing();
    performanceNotes();
    validationExample();
    runTests();

    console.log("\n" + "=".repeat(78));
    console.log("END OF SET OPERATIONS STUDY FILE");
    console.log("=".repeat(78));
}

main().catch((error) => {
    console.error("Program failed:", error);
    process.exitCode = 1;
});

File: day_27_set_operations/day_27_set_operations.py
"""
SET OPERATIONS
==============

A self-contained study file covering sets from absolute beginner through
advanced level, with executable examples, algorithms, validation, testing,
edge cases, performance considerations, and practical applications.

Run with:
    python set_operations.py
"""

from __future__ import annotations

from dataclasses import dataclass
from time import perf_counter
from typing import Any, Iterable


# ============================================================================
# 1. FUNDAMENTALS
# ============================================================================

def fundamentals() -> None:
    print("\n" + "=" * 78)
    print("1. FUNDAMENTALS")
    print("=" * 78)

    # A set stores unique hashable elements. Duplicate values disappear.
    numbers = {1, 2, 3, 3, 2, 1}
    print("Duplicates removed:", numbers)

    # An empty set must be created with set(), because {} creates a dictionary.
    empty_set = set()
    print("Empty set:", empty_set)

    # A set can contain numbers, strings, tuples, and other immutable/hashable
    # objects. Lists and dictionaries cannot directly be set elements.
    mixed = {10, "Python", (1, 2)}
    print("Hashable mixed set:", mixed)

    # Membership testing is a fundamental set operation.
    print("2 in numbers:", 2 in numbers)
    print("99 in numbers:", 99 in numbers)

    # Sets are unordered collections. Do not rely on printed order.
    # They do not support positional indexing.
    try:
        print(numbers[0])
    except TypeError as error:
        print("Indexing error:", error)

    # Cardinality means the number of elements in a set.
    print("Cardinality:", len(numbers))


# ============================================================================
# 2. CREATION AND CONVERSION
# ============================================================================

def creation_and_conversion() -> None:
    print("\n" + "=" * 78)
    print("2. CREATION AND CONVERSION")
    print("=" * 78)

    from_list = set([1, 2, 2, 4, 5])
    from_tuple = set((2, 3, 3, 6))
    from_string = set("banana")

    print("From list:", from_list)
    print("From tuple:", from_tuple)
    print("Unique characters:", from_string)

    # A dictionary converted to a set produces its keys.
    dictionary = {"id": 1, "name": "Atul"}
    print("Dictionary keys as set:", set(dictionary))

    # A frozenset is immutable and therefore hashable.
    immutable_set = frozenset({1, 2, 3})
    print("Frozenset:", immutable_set)

    # A frozenset can itself be an element of another set.
    nested_sets = {frozenset({1, 2}), frozenset({3, 4})}
    print("Set containing frozensets:", nested_sets)


# ============================================================================
# 3. ADDING AND REMOVING ELEMENTS
# ============================================================================

def mutation_operations() -> None:
    print("\n" + "=" * 78)
    print("3. ADDING AND REMOVING ELEMENTS")
    print("=" * 78)

    values = {1, 2, 3}

    values.add(4)
    print("After add:", values)

    values.update([5, 6, 6])
    print("After update:", values)

    # remove() raises KeyError when the element does not exist.
    values.remove(6)
    print("After remove:", values)

    # discard() silently does nothing when the element is absent.
    values.discard(999)
    print("After discard of missing value:", values)

    # pop() removes an arbitrary element. Never assume which element it removes.
    removed = values.pop()
    print("Arbitrarily removed:", removed)
    print("Remaining:", values)

    values.clear()
    print("After clear:", values)


# ============================================================================
# 4. MATHEMATICAL SET OPERATIONS
# ============================================================================

def mathematical_operations() -> None:
    print("\n" + "=" * 78)
    print("4. MATHEMATICAL SET OPERATIONS")
    print("=" * 78)

    A = {1, 2, 3, 4}
    B = {3, 4, 5, 6}

    # Union: elements belonging to A or B.
    print("Union:", A | B)
    print("Union method:", A.union(B))

    # Intersection: elements belonging to both A and B.
    print("Intersection:", A & B)
    print("Intersection method:", A.intersection(B))

    # Difference: elements in A that are not in B.
    print("A - B:", A - B)
    print("B - A:", B - A)

    # Symmetric difference: elements in exactly one of the two sets.
    print("Symmetric difference:", A ^ B)
    print("Symmetric difference method:", A.symmetric_difference(B))

    # Equivalent update operations mutate the left-hand set.
    C = A.copy()
    C |= B
    print("Union update:", C)

    C = A.copy()
    C &= B
    print("Intersection update:", C)

    C = A.copy()
    C -= B
    print("Difference update:", C)

    C = A.copy()
    C ^= B
    print("Symmetric-difference update:", C)


# ============================================================================
# 5. RELATIONS BETWEEN SETS
# ============================================================================

def set_relations() -> None:
    print("\n" + "=" * 78)
    print("5. RELATIONS BETWEEN SETS")
    print("=" * 78)

    A = {1, 2}
    B = {1, 2, 3}
    C = {4, 5}

    # Subset: every element of A is also in B.
    print("A <= B:", A <= B)
    print("A.issubset(B):", A.issubset(B))

    # Proper subset requires A != B.
    print("A < B:", A < B)

    # Superset is the reverse relationship.
    print("B >= A:", B >= A)
    print("B.issuperset(A):", B.issuperset(A))

    # Disjoint sets have no common elements.
    print("A and C disjoint:", A.isdisjoint(C))

    # Equal sets contain exactly the same elements regardless of insertion order.
    print("{1, 2} == {2, 1}:", {1, 2} == {2, 1})


# ============================================================================
# 6. SET COMPREHENSIONS
# ============================================================================

def set_comprehensions() -> None:
    print("\n" + "=" * 78)
    print("6. SET COMPREHENSIONS")
    print("=" * 78)

    # Set comprehension combines iteration, transformation, and uniqueness.
    squares = {number * number for number in range(1, 8)}
    print("Squares:", squares)

    even_squares = {
        number * number
        for number in range(1, 11)
        if number % 2 == 0
    }
    print("Even squares:", even_squares)

    words = ["apple", "Apple", "banana", "BANANA", "pear"]
    normalized_words = {word.lower() for word in words}
    print("Normalized unique words:", normalized_words)


# ============================================================================
# 7. REAL-WORLD EXAMPLE: USER ACCESS
# ============================================================================

def access_control_example() -> None:
    print("\n" + "=" * 78)
    print("7. REAL-WORLD EXAMPLE: ACCESS CONTROL")
    print("=" * 78)

    required_permissions = {"read", "write"}
    user_permissions = {"read", "write", "analytics"}

    has_required_access = required_permissions.issubset(user_permissions)
    extra_permissions = user_permissions - required_permissions
    missing_permissions = required_permissions - user_permissions

    print("Required permissions:", required_permissions)
    print("User permissions:", user_permissions)
    print("Authorized:", has_required_access)
    print("Extra permissions:", extra_permissions)
    print("Missing permissions:", missing_permissions)


# ============================================================================
# 8. REAL-WORLD EXAMPLE: DATA RECONCILIATION
# ============================================================================

def reconciliation_example() -> None:
    print("\n" + "=" * 78)
    print("8. REAL-WORLD EXAMPLE: DATA RECONCILIATION")
    print("=" * 78)

    database_ids = {101, 102, 103, 104, 105}
    api_ids = {103, 104, 105, 106, 107}

    common = database_ids & api_ids
    only_database = database_ids - api_ids
    only_api = api_ids - database_ids
    all_ids = database_ids | api_ids

    print("Common IDs:", common)
    print("Only database:", only_database)
    print("Only API:", only_api)
    print("Combined IDs:", all_ids)


# ============================================================================
# 9. FUNCTIONS THAT ACCEPT ITERABLES
# ============================================================================

def unique_preserving_order(values: Iterable[Any]) -> list[Any]:
    """
    Return unique values while preserving first-seen order.

    A plain set is excellent for uniqueness and membership, but its mathematical
    nature does not guarantee an application-level ordering contract.
    """
    seen: set[Any] = set()
    result: list[Any] = []

    for value in values:
        if value not in seen:
            seen.add(value)
            result.append(value)

    return result


def multiset_difference(left: Iterable[Any], right: Iterable[Any]) -> list[Any]:
    """
    Demonstrate a limitation of ordinary sets.

    A set loses multiplicity, so this function uses a dictionary of counts to
    perform a multiset-style difference.
    """
    from collections import Counter

    left_counts = Counter(left)
    right_counts = Counter(right)
    result: list[Any] = []

    for value, count in left_counts.items():
        remaining = count - right_counts[value]
        result.extend([value] * max(0, remaining))

    return result


def advanced_collection_examples() -> None:
    print("\n" + "=" * 78)
    print("9. ADVANCED COLLECTION EXAMPLES")
    print("=" * 78)

    values = [4, 2, 4, 1, 2, 3, 1]
    print("Input:", values)
    print("set(values):", set(values))
    print("Ordered unique values:", unique_preserving_order(values))

    left = ["a", "a", "a", "b", "c"]
    right = ["a", "b"]
    print("Multiset difference:", multiset_difference(left, right))


# ============================================================================
# 10. HASHABILITY AND EDGE CASES
# ============================================================================

def hashability_and_edge_cases() -> None:
    print("\n" + "=" * 78)
    print("10. HASHABILITY AND EDGE CASES")
    print("=" * 78)

    # Mutable containers such as lists are not hashable and cannot be elements
    # of a normal set.
    try:
        invalid = {[1, 2], [3, 4]}
        print(invalid)
    except TypeError as error:
        print("Unhashable element:", error)

    # Tuples are hashable when all of their members are hashable.
    valid = {(1, 2), (3, 4)}
    print("Tuple elements:", valid)

    # A tuple containing a list is not hashable.
    try:
        invalid_tuple = {(1, [2, 3])}
        print(invalid_tuple)
    except TypeError as error:
        print("Unhashable nested value:", error)

    # NaN has unusual equality behavior. It is a floating-point special value,
    # not an ordinary mathematical real number.
    nan = float("nan")
    nan_set = {nan}
    print("NaN membership using the same object:", nan in nan_set)

    # None is hashable and can be stored normally.
    print("None in set:", {None, 1, 2})


# ============================================================================
# 11. RELATION TO OTHER COLLECTIONS
# ============================================================================

def collection_comparison() -> None:
    print("\n" + "=" * 78)
    print("11. COLLECTION COMPARISON")
    print("=" * 78)

    values = [1, 2, 2, 3]
    tuple_values = (1, 2, 2, 3)
    set_values = {1, 2, 2, 3}
    dictionary = {"a": 1, "b": 2}

    print("List preserves duplicates and order:", values)
    print("Tuple preserves duplicates and order:", tuple_values)
    print("Set stores unique values:", set_values)
    print("Dictionary maps keys to values:", dictionary)

    print("List membership:", 3 in values)
    print("Set membership:", 3 in set_values)


# ============================================================================
# 12. ALGORITHMIC IMPLEMENTATIONS
# ============================================================================

def manual_union(first: set[Any], second: set[Any]) -> set[Any]:
    result = set(first)
    result.update(second)
    return result


def manual_intersection(first: set[Any], second: set[Any]) -> set[Any]:
    # Iterate through the smaller set to reduce membership checks.
    smaller, larger = (
        (first, second) if len(first) <= len(second) else (second, first)
    )
    return {value for value in smaller if value in larger}


def manual_difference(first: set[Any], second: set[Any]) -> set[Any]:
    return {value for value in first if value not in second}


def manual_symmetric_difference(
    first: set[Any], second: set[Any]
) -> set[Any]:
    return (first - second) | (second - first)


def algorithmic_examples() -> None:
    print("\n" + "=" * 78)
    print("12. ALGORITHMIC IMPLEMENTATIONS")
    print("=" * 78)

    A = {1, 2, 3, 4}
    B = {3, 4, 5, 6}

    print("Manual union:", manual_union(A, B))
    print("Manual intersection:", manual_intersection(A, B))
    print("Manual difference:", manual_difference(A, B))
    print("Manual symmetric difference:", manual_symmetric_difference(A, B))


# ============================================================================
# 13. INVERTED INDEX
# ============================================================================

def build_inverted_index(documents: dict[str, str]) -> dict[str, set[str]]:
    """
    Map each normalized word to the documents containing it.

    This is a simplified search-engine-style inverted index.
    """
    index: dict[str, set[str]] = {}

    for document_id, text in documents.items():
        words = {
            word.strip(".,!?;:()[]{}\"'").lower()
            for word in text.split()
        }

        for word in words:
            if word:
                index.setdefault(word, set()).add(document_id)

    return index


def search_documents(
    index: dict[str, set[str]],
    required_terms: Iterable[str],
) -> set[str]:
    terms = [term.lower() for term in required_terms]

    if not terms:
        return set()

    matching_documents = index.get(terms[0], set()).copy()

    for term in terms[1:]:
        matching_documents &= index.get(term, set())

    return matching_documents


def inverted_index_example() -> None:
    print("\n" + "=" * 78)
    print("13. INVERTED INDEX")
    print("=" * 78)

    documents = {
        "doc1": "Python supports set operations and dictionaries.",
        "doc2": "Set operations are useful for data processing.",
        "doc3": "Python data structures include lists sets and dictionaries.",
    }

    index = build_inverted_index(documents)

    print("Documents containing 'python':", index.get("python", set()))
    print(
        "Documents containing both 'set' and 'operations':",
        search_documents(index, ["set", "operations"]),
    )


# ============================================================================
# 14. GRAPH APPLICATION
# ============================================================================

def graph_neighbors_example() -> None:
    print("\n" + "=" * 78)
    print("14. GRAPH APPLICATION")
    print("=" * 78)

    graph = {
        "A": {"B", "C"},
        "B": {"A", "C", "D"},
        "C": {"A", "B"},
        "D": {"B"},
    }

    # Neighbor sets make operations such as common-neighbor detection simple.
    common_neighbors = graph["A"] & graph["B"]
    print("Common neighbors of A and B:", common_neighbors)

    # A set is also useful for tracking visited graph vertices.
    visited: set[str] = set()
    stack = ["A"]

    while stack:
        vertex = stack.pop()

        if vertex in visited:
            continue

        visited.add(vertex)
        stack.extend(graph[vertex] - visited)

    print("Visited vertices:", visited)


# ============================================================================
# 15. PERFORMANCE
# ============================================================================

def performance_example() -> None:
    print("\n" + "=" * 78)
    print("15. PERFORMANCE")
    print("=" * 78)

    size = 100_000
    values = list(range(size))
    value_set = set(values)
    target = size - 1

    start = perf_counter()
    _ = target in values
    list_time = perf_counter() - start

    start = perf_counter()
    _ = target in value_set
    set_time = perf_counter() - start

    print(f"Membership in list: {list_time:.8f} seconds")
    print(f"Membership in set:  {set_time:.8f} seconds")
    print(
        "Typical set membership is average O(1), while list membership is O(n)."
    )
    print(
        "Measured times depend on hardware, Python version, workload, and cache."
    )


# ============================================================================
# 16. TESTS
# ============================================================================

def run_tests() -> None:
    print("\n" + "=" * 78)
    print("16. SELF-TESTS")
    print("=" * 78)

    A = {1, 2, 3}
    B = {3, 4, 5}

    assert A | B == {1, 2, 3, 4, 5}
    assert A & B == {3}
    assert A - B == {1, 2}
    assert A ^ B == {1, 2, 4, 5}
    assert {1, 2}.issubset(A)
    assert A.issuperset({1, 2})
    assert A.isdisjoint({8, 9})
    assert unique_preserving_order([1, 2, 1, 3]) == [1, 2, 3]
    assert multiset_difference(["a", "a", "b"], ["a"]) == ["a", "b"]

    print("All tests passed.")


# ============================================================================
# 17. PRACTICAL DATA-QUALITY PIPELINE
# ============================================================================

@dataclass
class DataQualityReport:
    total_records: int
    unique_records: int
    duplicates: set[str]
    blank_records: set[str]


def analyze_records(records: Iterable[str]) -> DataQualityReport:
    normalized = [record.strip().lower() for record in records]
    unique_records = set(normalized)

    duplicates = {
        value for value in unique_records if normalized.count(value) > 1
    }
    blank_records = {value for value in unique_records if not value}

    return DataQualityReport(
        total_records=len(normalized),
        unique_records=len(unique_records),
        duplicates=duplicates,
        blank_records=blank_records,
    )


def data_quality_example() -> None:
    print("\n" + "=" * 78)
    print("17. DATA-QUALITY PIPELINE")
    print("=" * 78)

    records = [
        "Alice@example.com",
        "alice@example.com",
        " bob@example.com ",
        "",
        "carol@example.com",
        "BOB@EXAMPLE.COM",
    ]

    report = analyze_records(records)

    print("Total records:", report.total_records)
    print("Unique records:", report.unique_records)
    print("Duplicates:", report.duplicates)
    print("Blank values:", report.blank_records)


# ============================================================================
# 18. ADVANCED NOTES
# ============================================================================

def advanced_notes() -> None:
    print("\n" + "=" * 78)
    print("18. ADVANCED NOTES")
    print("=" * 78)

    print("Set elements must be hashable.")
    print("Hashable objects need stable hashing/equality behavior.")
    print("Mutating an object in a way that changes its hash while it is used")
    print("as a set element would violate the set's lookup assumptions.")
    print("frozenset is the immutable counterpart of set.")
    print("set operations normally provide average constant-time membership.")
    print("Worst-case hash-table behavior can degrade under pathological collisions.")
    print("Set ordering should not be treated as an application contract.")
    print("Sets discard duplicates and therefore cannot represent multiplicity.")
    print("For counts, use a frequency map or Counter.")
    print("For deterministic presentation, sort the resulting set explicitly.")


# ============================================================================
# 19. MAIN
# ============================================================================

def main() -> None:
    print("=" * 78)
    print("SET OPERATIONS: BEGINNER TO ADVANCED")
    print("=" * 78)

    fundamentals()
    creation_and_conversion()
    mutation_operations()
    mathematical_operations()
    set_relations()
    set_comprehensions()
    access_control_example()
    reconciliation_example()
    advanced_collection_examples()
    hashability_and_edge_cases()
    collection_comparison()
    algorithmic_examples()
    inverted_index_example()
    graph_neighbors_example()
    performance_example()
    data_quality_example()
    advanced_notes()
    run_tests()

    print("\n" + "=" * 78)
    print("END OF SET OPERATIONS STUDY FILE")
    print("=" * 78)


if __name__ == "__main__":
    main()
